### DISCUSSION TITLE
Conversion of energy counter sensors from Wh to kWh?

### DISCUSSION THREAD

**Discussion by noxhirsch** [SukramJ/aiohomematic#208] [Ideas]
Home Assistant expects energy counter sensors to be in `kWh`, but Homematic deliveres them in `Wh`. I think it would be useful if hahomematic would do the conversion job for the users, so they don't have to create custom sensors just to move the decimal point to the left.

<img width="329" alt="image" src="https://user-images.githubusercontent.com/30938717/149897783-2ba59ada-a1c9-41d8-b212-a4fa1943ee5f.png">

Daniel has already written this on the subject https://github.com/danielperna84/hahomematic/issues/158#issuecomment-1009767059 :
> After all there might be other cases where a transparent conversion might make sense. Like f.ex. the energy-sensors which are exposed as `Wh` from the devices, but the typical unit would be `kWh`. Hence the idea of at least a general framework which would allow doing such things in case it is needed.

I think there could be two approaches:
1. Convert everything to `kWh`: This is the cleanest solution for the users but maybe confuse some of them, if they compare the value with Homematic
2. Create two sensors:
    - "Devicename Energy Counter": `kWh` - enabled by default
    - "Devicename Energy Counter RAW / Wh / ...": `Wh` - **disabled by default**

**Comment by danielperna84** [comment]
In case of the `Wh` I wouldn't make the `kWh` the default though. As a HomeMatic user `Wh` is would you would expect. If this value was missing the users would report issue of a missing sensor. 😅 But then again the `kWh` sensor would be a duplicate technically. So I'm undecided right now if we really should do this. If we do this here, then more requests for other customized sensors will probably pop up, and we'd always have to argue why we do or do not implement something.

Maybe we should establish some general _rules_ which should be followed. In the case of the `ppm` it did make sense, as the unit in the device was in some way buggy, not matching practical values. But for `Wh` and `kWh` it would be more of a convenince-thing, which for users who really need it can be worked around. After all the `Wh` sensors work for the energy dashboard (this didn't work when the energy-stuff was new, but now it does), so we don't _need_ this.

Other opinions of course are welcome.

**Reply by noxhirsch** [reply]
Okay, that actually works - I assume HomeAssistant will then take care of the conversion. Then actually only the label of the dialog (see screenshot above) is wrong. I'd also say that the conversion to hahomematic is unnecessary then. The only reason for my suggestion was because of the Energy Dashboard, which is probably used by many users. If you need kWh for other specific reasons, you can really create a custom sensor yourself.

**Reply by danielperna84** [reply]
The dialog is native Home Assistant. They won't change that. The list simply holds all compatible devices. It doesn't really matter if it's `kWh` or `Wh` in this case. So I guess this is only relevant if the users _knows_ his sensors provide the _wrong_ unit in regards to what the UI is saying. The only real optimization that could be done would be to change the `(kWh)` to `(kWh / Wh)`. But that's really just cosmetic for the "Erbsenzähler". 😜

**Reply by noxhirsch** [reply]
>The only real optimization that could be done would be to change the (kWh) to (kWh / Wh)

That's what I meant 👍 I'll open an issue or PR in the frontend repo for that, soon.



>But that's really just cosmetic for the "Erbsenzähler". 😜

I don't know who you're talking about 😜
